README.md
# Snake

![Snake Overview](images/overview.png  "Overview")

Snake is application to control and configure your Razer devices on Linux. Making use of the awesome OpenRazer drivers, *Snake* provides access to most features the drivers provides and lots of extra stuff you won't find in other clients.

 * Stylised user interface with animations and effects
 * Simple installation, and self updating.
 * Provides basic support from all devices supported by OpenRazer.
 * Layouts feature. Design and use labelled diagrams of your device showing lighting areas, keys and individual LEDs. If a device has a layout, it is used for controls, macros and custom effects instead of the basic generated user interface.
 * Supports multiple concurrent devices.
 * Set and configure Razer effects such as Breath, Wave, Static and more.
 * Create custom effects using the timeline based editor. Configurable with different interpolations types, FPS and speed.
 * Battery status.
 * Audio analysis lighting effects. Monitors a channel PulseAudio and makes your lighting react in different ways.
   Requires a device that has Matrix support.
 * Integrated System Tray icon (or indicator)
 * Global brightness controls the brightness of all devices at once.
 * Macro support. *(work in progress, 3 competing systems!)*
 * Themes
  
![A device with a layout](images/device-using-layout.png  "Device With Layout")

## Compatibility

Snake tries to be completely compatible with whatever OpenRazer supports. However, there are some features OpenRazer does not yet provide itself, snake such as detailed lighting and key layouts.

Snake includes a very small set of built in layouts, only for hardware I have access to. If you create a high quality layout for your device, please submit it to the project and I will add it as a default.

| Device                      | Macros  | Layouts           | 
| --------------------------- | --------| ----------------- |
| *Keyboards*                      |   |            | 
| Other OpenRazer Keyboards | 1,2   | None            |
| *Mice*                      |   |            |
|  <img src="https://assets.razerzone.com/eeimages/support/products/1390/1390_mamba_elite.png" width="64" />Mamba Elite                 | 3   | MATRIX, TOP       |
| Other OpenRazer Mice |  | None            | None            | 
| *Accessories*                      |   |            |
| <img src="https://assets.razerzone.com/eeimages/support/products/1255/1255_tartarus_v2.png" width="64" /> Tartarus V2				  | 3   | MATRIX,TOP        | 26              |
| <img src="https://assets2.razerzone.com/images/pnx.assets/132a2c378c62416d4d0634306f04bf01/connectivity-nommo-pro.jpg" width="64" /> Nommo Chroma                | 3   | MATRIX, THREE_D_1 | 2               |
| Other OpenRazer Accessories |   | None            | None            | 

1. OpenRazer default macro/key binding system. Only supports keyboards. Works at X11 / device driver levels. Works without a layout.
2. Chabad360's OpenRazer macro/key binding improvements  [https://github.com/chabad360/openrazer/tree/keybinding](branch). Currently only supports keyboards. Requires UInput permissions. Works without a layout.
3. Snakes own key binding system. Supports all device types. Requires UInput permissions. Requires a layout.
 
## Install

There are several different ways to run Snake. In all cases, the first time the app runs you will be asked if you want to create a desktop shortcut (i.e. menu entry).

 ![Snake Installer](images/install.png  "Snake Installer")

**All binary versions of Snake are built for 64 bit x86 only. For 32 bit, or other architectures, please use one of the "No-Runtime" packages below.**

### Self Extracting Installer

The self extracting installer is the easiest to use. You will be asked for a location to install Snake. It will then update to the latest version and start up.

#### From The Command Line

```
wget -O snake-installer-amd64.sh \
 https://github.com/bithatch/snake/releases/download/v1.0-SNAPSHOT-204/snake-installer-amd64.sh && \
 sh snake-installer-amd64.sh
```

#### From Your Desktop

 *  Download [snake-amd64-installer.sh](https://github.com/bithatch/snake/releases/download/v1.0-SNAPSHOT-204/snake-installer-amd64.sh).
 *  Add Execute permission to the downloaded *snake-installer.sh* in your file manager.
 *  Use the Open action in your file manager to start the installer.

### Application Archive

Another option is to download the archived version of the app, extract it to where you want to install, and just run it.

#### From The Command Line

```
wget -O snake-amd64.tar.gz \
 https://github.com/bithatch/snake/releases/download/v1.0-SNAPSHOT-204/snake-amd64.tar.gz
tar xzf snake-amd64-tar.gz
cd snake
bin/snake &
```

#### From Your Desktop

 *  Download [snake-amd64.tar.gz](https://github.com/bithatch/snake/releases/download/v1.0-SNAPSHOT-204/snake-amd64.tar.gz)
 *  Use the Open action in your file manager to locate and start `bin/snake` in the extracted folder.
 
### No-Runtime Packages

The above packages all include a stripped down amd64 Java runtime so you do not need to install your own. If you  have another architecture such as X86 or Arm, or just want a smaller download, you can use your own runtime. It just needs to be at least [Java 15](https://adoptopenjdk.net/). 

 * [Installer](https://github.com/bithatch/snake/releases/download/v1.0-SNAPSHOT-204/snake-installer-noruntime.sh)
 * [Tar Archive](https://github.com/bithatch/snake/releases/download/v1.0-SNAPSHOT-204/snake-noruntime.tar.gz)
 
All these packages include the `bin/snake` launcher script. You may need to edit this if you want to use a runtime other than the system default.
 
### OS Packages

Not currently available.

## Updating

All installations of Snake are self updating, unless you are using our own OS specific packages. Those provided by a 3rd party repository may provide updates.

Update options may be found may be found in the ** Options** screen.

### Automatic Updates

By default, Snake will update automatically and you should never need to touch it. If you prefer though, you can turn this off and you will be notified when updates are available. The update may then be installed on demand. 

### Beta Updates

You can optionally accept beta updates. While I will strive to keep the beta channel stable, these may contain as-yet untested features, and some behaviour may be unexpected. You can at any time revert to the release  channel.

When changing this option, *Snake* will restart after 2 seconds and download appropriate version.
 
## Build From Source

To build Snake from source you need :-

 * Git
 * Java 15+
 * Apache Maven

Other dependencies are downloaded by Maven during the build process.

Further dependencies may be needed for building deployable packages.

### Build

```
git clone https://github.com/bithatch/snake.git
cd snake
mvn clean install
```

If the build completes OK, you can then run or install Snake.

### Run

Once build, you can run directly.

```
cd snake-app-linux
mvn javafx:run
```

### Install

Or you can install what you have built.

```
snake-dist/target/snake-installer-amd64.sh
```

## Options

Configuration is very simple. The backend should take care of the devices, Snake just has a few options for appearance and behaviour.

 ![Snake Options](images/options.png  "Snake Options")

To access options, click the cog icon, ** Options**.

### Appearance

Various options for configuration how Snake looks.

**Theme** - Allows the look of Snake to be changed. A selection of themes that are colour variations of the default are provided, with more to come. Creating your own will be documented soon, but the adventurous can try by copying any existing theme module and adjusting the contents accordingly.

![Themes](images/themes.png  "Themes")

**Use Window Manager Decorations** - By default, Snake uses "Client Side Decorations", and draws it's own window title bar and window management buttons (close, min, max). Turn this option off to use window manager decorations instead. In this mode, Options and About are both insteads accessed from icons at the bottom right of the overview screen.

**Transparency** - If you are using client side decorations, then window transparency is possible. This slider sets how transparent the window should be.

![Snake Desktop](images/desktop.png  "Snake Desktop")

### Updates

**Check for updates** - Whether to check for updates at all. When deselected, neither automatic or manual updating is possible.

 **Update automatically** - When selected, Snake will update itself automatically on startup if required. 
 
 **Install beta updates as well** - When selected, you will be able to receive beta updates  that may contain fixes or features that you wish to try out before full release.

### Tray

**Icon  Style** - Different options for how the tray icon appears, if at all.

**Show Battery** - If you have at least one device that has a battery, then the option is given to show the battery status on the system tray instead of the razer logo.

  **When low** - When selected the battery status will only be showing if the overall battery status is *Low*.
  
**Start on login** - If the tray icon is enabled, you probably want to start it when you login. The option allows you to do so. When turned off, the tray icon will appear when you manually start Snake.

## System Tray

The system tray icon runs in the background and provides quick access to common functions of your devices.

 ![Snake Tray](images/tray.png  "Snake Tray")
 
 It can also show a summary of the battery status (when configuration).

## Device Overview

Device Overview is the first screen you see when opening Snake. It lists all of the supported Razer devices that the driver has discovered. 

If an expected device is not listed, please check if it works with other OpenRazer clients. If it does not, it is likely you want to direct any feature requests to the OpenRazer project.

If you have lots of devices, you can use the search and filter options to narrow down what is shown, although this is most useful when testing Snake with lots of 'Fake Drivers' that OpenRazer can supply.

Either double click anywhere on the row or click the Slider icon on the far right of each row to see the details for that device.

## Device Details

Double clicking on a device in the over page will show the configuration screen for that device.

It will show all of the supported *Controls*, depending on what device is in use.

 ![Device Details](images/device.png  "Device Details")

For example, controls are provided for :-

 * Brightness. Will allow configuration of both overall brightness, and per region brightness.
 * Effects. Choose and configure various standard Razer effects.
 * Poll Rate. Low level setting for mice.
 * DPI. Dots per inch setting, changes the accuracy of your mouse movements.
 * Battery. Shows battery status and allows configuration of low-battery thresholds.
 * Game Mode.  Turn game mode on or off. When on, the Windows key will be disabled.
 
 At any time, clicking on the *Back* icon at the top right of the screen will return you to the *Device Overview* screen.
 
### Brightness

Controls the intensity of  all lights in a region, or overall on the device. If the device only supports a single lighting region, only a single slider will be shown.

 ![Brightness](images/brightness.png  "Brightness")
 
### Poll Rate

A low level setting that determines the speed at which your computer communicates with this device.

 ![Poll Rate](images/pollrate.png  "Poll Rate")
 
### DPI

Dots per inch setting, changes the accuracy of your mouse movements.

The maximum setting here can differ between devices. The higher the number, the faster the mouse pointer will move across the screen as you physically move the house. A lower number means the mouse pointer will move slower across the screen with more accuracy.

A value of 1000 to 2000 is probably comfortable for daily usage.

 ![DPI](images/dpi.png  "DPI")
 
### Effects
 
Many devices are capable of multiple effects. Some of those effects have their own configuration, and often it is possible to configure different effects for each region.
 
#### None
 
 This effect type simply turns off all lighting. There is no further configuration.
 
#### Static

The simplest effect,  just sets a fixed colour either on the whole device or individual regions. 

 ![Static Effect](images/static.png  "Static")
 
#### Wave

The region or device will show a multi-colours moving wave, in one of two directions. 

 ![Wave Effect](images/wave.png  "Wave")
 
#### Spectrum

The region or device will cycle at a fixed speed through all colours. 

#### Reactive

The region or device will light up on key or button clicks or mouse wheel movement.

Configuration allows changing of the colour and sensitivity of these reactions.

 ![Reactive Effect](images/reactive.png  "Reactive")

#### Breath

The region or device will gently pulsate different colours.

Configuration allows changing of the colours and a special random mode.

 ![Breath](images/breath.png  "Breath")

#### Ripple

The region or device will light a ripple effect moving outwards from the centre of the device, with either a random colour or a fixed single colour.

 ![Ripple](images/ripple.png  "Ripple")

#### Pulsate

The region or device will gently pulsate a single fixed colour.

#### Starlight

The region or device will twinkle lights on and off using  either a random colour or a fixed single colour to mimick a night sky.

 ![Starlight](images/starlight.png  "Starlight")
 
#### Matrix

The Matrix effect may only be set on the whole device, and allows each LED to be individually configured.

The exact layout of the screen will depend on the device. If it is a keyboard, a virtual keyboard will be shown, otherwise just a row of numbered LED lights will be seen.

 ![Matrix](images/matrix.png  "Matrix")

## Macros

The Macro UI is available, but due to coming changes in OpenRazer this is likely to radically change, so I won't go into any detail here.

 ![Macros](images/macros.png  "Macros")

### Recording Macros

** Note, Recording functionality is not complete, but will be before final release. You can manually edit macros though.  **

 ![Record](images/record.png  "Record")


CHANGES.md
# 1.0.0-SNAPSHOT-${buildNumber}

 * Added layout for Razer Base Station Chroma V2 .
 * Updated to Java 16. This lets us use dbus-java 4.0.0-SNAPSHOT.
 * Updated to dbus-java 4.0.0-SNAPSHOT. This lets us drop a lot of dependencies.
 * Added an uninstaller (run bin/uninstaller)

# 1.0.0-SNAPSHOT-173

 * Fails to start if `jscal` or `jstest-gtk` are not installed.
   
# 1.0.0-SNAPSHOT-172

 * OpenJ9 wasn't actually activated (due to problem with jimpulse), but it now is.

# 1.0.0-SNAPSHOT-151

 * Switch to OpenJ9 for lower memory usage and faster startup time. 
 
 * Updated Dorkbox SystemTray to latest.
 
 * Added splash screen.

# 1.0.0-SNAPSHOT-148

 * Testing repository location accidentally baked into build. You will need to re-install or use 

   ```
   bin/snake --remote-manifest http://www.bithatch.co.uk/repositories/snake/snapshot/manifest.xml
   ```

   .. or if you are using the 'no-runtime' package ..
      
   ```
   bin/snake --remote-manifest http://www.bithatch.co.uk/repositories/snake-noruntime/snapshot/manifest.xml
   ```
   
# 1.0.0-SNAPSHOT-145
 
 * Audio lighting effects. Requires `libpulse` and `libfftw-3` to be installed. 

 * A 3rd  Macro/remapping implementation, this time based on my own MacroLib, a port of Gnome15's
   macro system to Java. Automatic profile selection is supported, but you will need `bamfdaemon` 
   installed (there is a fallback planned that uses libwnck3 but this is currently faulty).
   
 * Support hot-plugging devices.
 
 * Failed to start when "Start on login" selected.
 
 * Fixed some memory leaks.
 
 * Issues with switching between effects. Matrix effects especially could not be stacked.
 
 * Some UI improvements. "Controls" stand out better against the background. More consistent
   spacing. 
   
 * Incremental custom effects improvements. Each cell in a keyframe can now take either it's
   Hue, Saturation or Brightness from a static colour, the audio level or a random value. 
   This opens up a few more possibilities for custom effects. More to come in future releases.
   
 * New colour picker (http://llbit.se/?p=3331)[LuxColorPicker].
 
 * Due to font conflict when FontAwesome is installed locally or system-wide, icon usage 
   completely reworked.
 
 * Lots of internal stuff. Better modularisation of a number of application and 3rd party libraries
   (ongoing, will ultimately lead to a lighter install). Much smaller command line used to
   launch the app and update (easier to read in `ps` command etc).

# 1.0.0-SNAPSHOT-24

 * Better support for Tartarus V2. 
 
 * Partial implementation of new macro system (https://github.com/openrazer/openrazer/pull/1124).
   More to come.
   

# 1.0.0-SNAPSHOT-20

 * Layout Editor. Can be used to create graphical layouts of Razer devices for a more 
   interesting user interface. These layouts may then be exported and shared with others
   (including to the Snake project for inclusions). The intention is over time, all devices
   have their own layout. The layouts are then used in all other aspects of Snake, such
   as the main UI, the custom effects editor and the macro user interface.
   
 * New main user interface making use of layouts (if your device has one). 
      
 * You can now create custom effects on devices that have the Matrix capability. Animations
   are based around the concept of Key Frames. All lights in the matrix interpolate between
   the current key frame and the next (using a configurable algorithm). The effects can
   be exported the be shared with others as an add-on.
   
 * Smoother UI effects.
 
 * Lots of internal refactoring. 

# 1.0.0-SNAPSHOT-1

 * Saner version numbers
 * New original application and tray icon
 * Cope with OpenRazer DBUS API version differences with getDeviceImage.
 * When opening from the tray, navigation could get messed up.

# 1.0.0-SNAPSHOT-0

Initial release of *Snake*.


snake-ui/src/main/java/uk/co/bithatch/snake/ui/Tray.java
package uk.co.bithatch.snake.ui;

import java.awt.AlphaComposite;
import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.ResourceBundle;

import javax.swing.JMenu;
import javax.swing.JSeparator;
import javax.swing.SwingUtilities;
import javax.swing.Timer;

import dorkbox.systemTray.Checkbox;
import dorkbox.systemTray.Entry;
import dorkbox.systemTray.Menu;
import dorkbox.systemTray.MenuItem;
import dorkbox.systemTray.Separator;
import dorkbox.systemTray.SystemTray;
import javafx.application.Platform;
import uk.co.bithatch.snake.lib.Backend.BackendListener;
import uk.co.bithatch.snake.lib.Capability;
import uk.co.bithatch.snake.lib.Device;
import uk.co.bithatch.snake.lib.Device.Listener;
import uk.co.bithatch.snake.lib.binding.Profile;
import uk.co.bithatch.snake.lib.binding.ProfileMap;
import uk.co.bithatch.snake.lib.Grouping;
import uk.co.bithatch.snake.lib.Item;
import uk.co.bithatch.snake.lib.Lit;
import uk.co.bithatch.snake.lib.Region;
import uk.co.bithatch.snake.ui.Configuration.TrayIcon;
import uk.co.bithatch.snake.ui.effects.EffectAcquisition;
import uk.co.bithatch.snake.ui.widgets.Direction;

public class Tray implements AutoCloseable, BackendListener, Listener {

	final static ResourceBundle bundle = ResourceBundle.getBundle(Tray.class.getName());

	private Configuration cfg;
	private App context;
	private Font font;
	private SystemTray systemTray;
	private List<Entry> menuEntries = new ArrayList<>();

	private Timer timer;

	public Tray(App context) throws Exception {
		this.context = context;

		font = Font.createFont(Font.TRUETYPE_FONT, Tray.class.getResourceAsStream("fontawesome-webfont.ttf"));

		cfg = context.getConfiguration();
		cfg.themeProperty().addListener((e) -> SwingUtilities.invokeLater(() -> adjustTray()));
		cfg.trayIconProperty().addListener((e) -> SwingUtilities.invokeLater(() -> adjustTray()));
		cfg.showBatteryProperty().addListener((e) -> SwingUtilities.invokeLater(() -> adjustTray()));
		cfg.whenLowProperty().addListener((e) -> SwingUtilities.invokeLater(() -> adjustTray()));

		context.getBackend().addListener(this);
		for (Device dev : context.getBackend().getDevices()) {
			dev.addListener(this);
		}
		SwingUtilities.invokeLater(() -> adjustTray());

	}

	@Override
	public void close() throws Exception {
		if (systemTray != null) {
			systemTray.shutdown();
			systemTray = null;
		}
	}

	Menu addDevice(Device device, Menu toMenu) throws IOException {
		var img = context.getDefaultImage(device.getType(), context.getCache().getCachedImage(device.getImage()));
		if (img.startsWith("http:") || img.startsWith("https:")) {
			var url = new URL(img);
			var path = url.getPath();
			var idx = path.lastIndexOf('/');
			if (idx != -1) {
				path = path.substring(idx + 1);
			}
			var tf = File.createTempFile("snake", path);
			tf.deleteOnExit();
			try (var fos = new FileOutputStream(tf)) {
				try (InputStream is = url.openStream()) {
					is.transferTo(fos);
				}
			}
			img = tf.getAbsolutePath();
		} else if (img.startsWith("file:")) {
			img = img.substring(5);
		}
		var menu = toMenu == null ? new Menu(device.getName(), img) : toMenu;
		var openDev = new MenuItem(bundle.getString("open"), (e) -> Platform.runLater(() -> {
			try {
				context.open();
				context.openDevice(device);
			} catch (Exception ex) {
			}
		}));
		menu.add(openDev);
		menuEntries.add(openDev);
		if (device.getCapabilities().contains(Capability.MACROS)) {
			var macros = new MenuItem(bundle.getString("macros"), createAwesomeIcon(bundle.getString("macrosIcon"), 32),
					(e) -> {
						Platform.runLater(() -> {
							context.open();
							context.push(Macros.class, Direction.FROM_BOTTOM).setDevice(device);
						});
					});
			menu.add(macros);
			menuEntries.add(macros);
		}
		if (device.getCapabilities().contains(Capability.BRIGHTNESS)) {
			brightnessMenu(menu, device);
		}
		if (device.getCapabilities().contains(Capability.DPI)) {
			dpiMenu(menu, device);
		}
		if (device.getCapabilities().contains(Capability.POLL_RATE)) {
			pollRateMenu(menu, device);
		}
		if (device.getCapabilities().contains(Capability.GAME_MODE)) {
			gameMode(menu, device);
		}
		menu.add(new JSeparator());
		if (!device.getSupportedEffects().isEmpty()) {
			addEffectsMenu(device, menu, device, toMenu != null);
			menu.add(new JSeparator());
		}
		boolean sep;
		for (var r : device.getRegions()) {
			var mi = new Menu(bundle.getString("region." + r.getName().name()));
			mi.setImage(context.getConfiguration().themeProperty().getValue().getRegionImage(24, r.getName()));
			menu.add(mi);
			if (toMenu != null)
				menuEntries.add(mi);
			sep = false;
			if (r.getCapabilities().contains(Capability.BRIGHTNESS_PER_REGION)) {
				brightnessMenu(mi, r);
				sep = true;
			}
			if (r.getCapabilities().contains(Capability.EFFECT_PER_REGION)) {
				if (sep) {
					mi.add(new JSeparator());
				}
				addEffectsMenu(r, mi, device, toMenu != null);
			}
		}
		return menu;
	}

	void addEffectsMenu(Lit lit, Menu menu, Device device, boolean addToRoot) {
		EffectHandler<?, ?> configurable = null;
		EffectAcquisition acq = context.getEffectManager().getRootAcquisition(Lit.getDevice(lit));
		EffectHandler<?, ?> selected = acq.getEffect(lit);
		for (EffectHandler<?, ?> fx : context.getEffectManager().getEffects(lit)) {
			var mi = new MenuItem(fx.getDisplayName(), (e) -> {
				context.getScheduler().execute(() -> acq.activate(lit, fx));
			});
			mi.setImage(fx.getEffectImage(24));
			menu.add(mi);
			if (fx.hasOptions() && fx.equals(selected)) {
				configurable = fx;
			}
			if (addToRoot)
				menuEntries.add(mi);
		}
		EffectHandler<?, ?> fConfigurable = configurable;
		if (configurable != null) {
			var mi = new MenuItem(bundle.getString("effectOptions"),
					createAwesomeIcon(bundle.getString("effectOptionsIcon"), 32), (e) -> {
						Platform.runLater(() -> {
							context.open();
							var dev = context.push(DeviceDetails.class, Direction.FROM_RIGHT);
							dev.setDevice(device);
							dev.configure(lit, fConfigurable);
						});
					});
			menu.add(mi);
			if (addToRoot)
				menuEntries.add(mi);
		}
	}

	BufferedImage createAwesomeIcon(String string, int sz) {
		return createAwesomeIcon(string, sz, 100);
	}

	BufferedImage createAwesomeIcon(String string, int sz, int opac) {
		return createAwesomeIcon(string, sz, opac, null, 0);
	}

	BufferedImage createAwesomeIcon(String string, int sz, int opac, Color col, double rot) {
		var bim = new BufferedImage(sz, sz, BufferedImage.TYPE_INT_ARGB);
		var graphics = (Graphics2D) bim.getGraphics();
		graphics.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
		var szf = (int) ((float) sz * 0.75f);
		var fnt = font.deriveFont((float) szf);
		graphics.setFont(fnt);
		if (opac != 100) {
			graphics.setComposite(makeComposite((float) opac / 100.0f));
		}
		var icon = cfg.trayIconProperty().getValue();
		if (col == null) {
			switch (icon) {
			case LIGHT:
				graphics.setColor(Color.WHITE);
				break;
			case DARK:
				graphics.setColor(Color.BLACK);
				break;
			default:
				// TODO
				break;
			}
		} else {
			graphics.setColor(col);
		}
		graphics.translate(sz / 2, sz / 2);
		graphics.rotate(Math.toRadians(rot));
		graphics.translate(-(graphics.getFontMetrics().stringWidth(string) / 2f),
				(graphics.getFontMetrics().getHeight() / 2f));
		graphics.drawString(string, 0, 0);
		return bim;
	}

	AlphaComposite makeComposite(float alpha) {
		int type = AlphaComposite.SRC_OVER;
		return (AlphaComposite.getInstance(type, alpha));
	}

	void adjustTray() {
		var icon = cfg.trayIconProperty().getValue();
		if (systemTray == null && icon != TrayIcon.OFF) {
			systemTray = SystemTray.get();
			if (systemTray == null) {
				throw new RuntimeException("Unable to load SystemTray!");
			}

			setImage();

			systemTray.setStatus(bundle.getString("title"));
			Menu menu = systemTray.getMenu();
			if (menu == null) {
				systemTray.setMenu(new JMenu(bundle.getString("title")));
			}

			rebuildMenu();

		} else if (systemTray != null && icon == TrayIcon.OFF) {
			systemTray.setEnabled(false);
		} else if (systemTray != null) {
			systemTray.setEnabled(true);
			setImage();
			rebuildMenu();
		}
	}

	private void rebuildMenu() {
		clearMenus();
		if (systemTray != null) {
			var menu = systemTray.getMenu();
			try {

				boolean devices = false;
				List<Device> devs = context.getBackend().getDevices();
				var backend = context.getBackend();
				var globals = false;

				if (devs.size() == 1) {
					addDevice(devs.get(0), menu);
					devices = true;
				} else {
					var mi = new MenuItem(bundle.getString("open"), (e) -> context.open());
					menuEntries.add(mi);
					menu.add(mi);
					addSeparator(menu);

					if (backend.getCapabilities().contains(Capability.BRIGHTNESS)) {
						brightnessMenu(menu, backend);
						globals = true;
					}
					if (backend.getCapabilities().contains(Capability.GAME_MODE)) {
						gameMode(menu, backend);
						globals = true;
					}
					if (globals)
						addSeparator(menu);
					for (Device dev : devs) {
						var devmenu = addDevice(dev, null);
						systemTray.getMenu().add(devmenu);
						menuEntries.add(devmenu);
						devices = true;
					}
				}
				if (devices)
					addSeparator(menu);
			} catch (Exception e) {
				// TODO add error item / tooltip?
				systemTray.setTooltip("Erro!");
				e.printStackTrace();
			}

			var quit = new MenuItem(bundle.getString("quit"), (e) -> context.close(true));
			menuEntries.add(quit);
			menu.add(quit).setShortcut('q');
		}
	}

	private void addSeparator(Menu menu) {
		Separator sep = new Separator();
		menu.add(sep);
		menuEntries.add(sep);

	}

	private void clearMenus() {
		if (systemTray != null) {
			var menu = systemTray.getMenu();
			for (Entry dev : menuEntries) {
				menu.remove(dev);
			}
		}
		menuEntries.clear();
	}

	private void gameMode(Menu menu, Grouping backend) {

		Checkbox checkbox = new Checkbox();
		checkbox.setCallback((e) -> backend.setGameMode(checkbox.getChecked()));
		checkbox.setEnabled(true);
		checkbox.setChecked(backend.isGameMode());
		checkbox.setShortcut('g');
		checkbox.setText(bundle.getString("gameMode"));

		menu.add(checkbox);
		menuEntries.add(checkbox);
	}

	private void brightnessMenu(Menu menu, Item backend) {
		var bri = new Menu(bundle.getString("brightness"), createAwesomeIcon(bundle.getString("brightnessIcon"), 32));
		menu.add(bri);
		for (String b : new String[] { "off", "10", "20", "50", "75", "100" }) {
			bri.add(new MenuItem(bundle.getString("brightness." + b),
					createAwesomeIcon(
							b.equals("off") ? bundle.getString("offIcon") : bundle.getString("brightnessIcon"), 32,
							b.equals("off") ? 25 : Integer.parseInt(b)),
					(e) -> {
						backend.setBrightness(b.equals("off") ? 0 : Short.parseShort(b));
					}));
		}
		menuEntries.add(bri);
	}

	private void dpiMenu(Menu menu, Device device) {
		var dpiMenu = new Menu(bundle.getString("dpi"), createAwesomeIcon(bundle.getString("dpiIcon"), 32));
		menu.add(dpiMenu);
		short[] dpis = new short[] { 100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 1500, 2000, 3000, 4000, 8000,
				16000, 32000 };
		for (short dpi : dpis) {
			if (dpi <= device.getMaxDPI()) {
				dpiMenu.add(new MenuItem(MessageFormat.format(bundle.getString("deviceDpi"), dpi), (e) -> {
					device.setDPI(dpi, dpi);
				}));
			}
		}
		menuEntries.add(dpiMenu);
	}

	private void pollRateMenu(Menu menu, Device device) {
		var pollRateMenu = new Menu(bundle.getString("pollRate"),
				createAwesomeIcon(bundle.getString("pollRateIcon"), 32));
		menu.add(pollRateMenu);
		int[] pollRates = new int[] { 125, 500, 1000 };
		for (int pollRate : pollRates) {
			pollRateMenu.add(new MenuItem(MessageFormat.format(bundle.getString("devicePollRate"), pollRate), (e) -> {
				device.setPollRate(pollRate);
			}));
		}
		menuEntries.add(pollRateMenu);
	}

	private void setImage() {
		var icon = cfg.trayIconProperty().getValue();
		if (context.getBackend().getCapabilities().contains(Capability.BATTERY) && cfg.showBatteryProperty().getValue()
				&& (!cfg.whenLowProperty().getValue() || (cfg.whenLowProperty().getValue()
						&& context.getBackend().getBattery() <= context.getBackend().getLowBatteryThreshold()))) {
			Color col = null;
			String status = BatteryControl.getBatteryStyle(context.getBackend().getLowBatteryThreshold(),
					context.getBackend().getBattery());
			if ("danger".equals(status))
				col = Color.red;
			else if ("warning".equals(status))
				col = Color.orange;
			systemTray.setImage(createAwesomeIcon(BatteryControl.getBatteryIcon(context.getBackend().getBattery()), 32,
					100, col, 270));
		} else {
			switch (icon) {
			case LIGHT:
				systemTray.setImage(App.class.getResource("icons/tray-light32.png"));
				break;
			case DARK:
				systemTray.setImage(App.class.getResource("icons/tray-dark32.png"));
				break;
			case AUTO:
				// TODO
			case COLOR:
			default:
				systemTray.setImage(
						context.getConfiguration().themeProperty().getValue().getResource("icons/tray32.png"));
				break;
			}
		}
	}

	@Override
	public void deviceAdded(Device device) {
		device.addListener(this);
		SwingUtilities.invokeLater(() -> rebuildMenu());
	}

	@Override
	public void deviceRemoved(Device device) {
		device.removeListener(this);
		SwingUtilities.invokeLater(() -> rebuildMenu());
	}

	@Override
	public void changed(Device device, Region region) {
		SwingUtilities.invokeLater(() -> {
			if (timer == null) {
				timer = new Timer(500, (e) -> {
					rebuildMenu();
				});
				timer.setRepeats(false);
			}
			timer.restart();
		});
	}

	@Override
	public void activeMapChanged(ProfileMap map) {
	}

	@Override
	public void profileAdded(Profile profile) {
	}

	@Override
	public void profileRemoved(Profile profile) {
	}

	@Override
	public void mapAdded(ProfileMap profile) {
	}

	@Override
	public void mapChanged(ProfileMap profile) {
	}

	@Override
	public void mapRemoved(ProfileMap profile) {
	}
}
